File: python/303-Range Sum Query - Immutable.py
from typing import List

# class NumArray:
#     def __init__(self, nums: List[int]):
#         "Precompute, Time: O(n^2), Space: O(n^2), TLE"
#         n = len(nums)
#         dp = [[0]*n for _ in range(n)] # dp[i][j]: sum of [i, j]
#         for i in range(n):
#             for j in range(i, n):
#                 if i == j: 
#                     dp[i][j] = nums[i]
#                 else:
#                     dp[i][j] = dp[i][j-1] + nums[j]
#         self.dp = dp
#     def sumRange(self, i: int, j: int) -> int:
#         "Time: O(1)"
#         return self.dp[i][j]

class NumArray:
    def __init__(self, nums: List[int]):
        "Precompute, Time: O(n), Space: O(n)"
        n = len(nums)
        if n == 0:
            return
        dp = [0]*n # dp[i] = sum of [0, i]
        dp[0] = nums[0]
        for i in range(1, n):
            dp[i] = dp[i-1] + nums[i]
        self.dp = dp
    def sumRange(self, i: int, j: int) -> int:
        "Time: O(1)"
        "(i, j) = (0, j) - (0, i-1)"
        if self.dp is None: return 0
        if i == 0:
            return self.dp[j]
        return self.dp[j] - self.dp[i-1]
        
numArray = NumArray([-2, 0, 3, -5, 2, -1])
print(numArray.sumRange(0, 2)) # return 1 ((-2) + 0 + 3)
print(numArray.sumRange(2, 5)) # return -1 (3 + (-5) + 2 + (-1)) 
print(numArray.sumRange(0, 5)) # return -3 ((-2) + 0 + 3 + (-5) + 2 + (-1))

File: cpp/303. Range Sum Query - Immutable.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class NumArray {
public:
    vi nums;
    vi prefixSum;
    NumArray(vector<int>& nums) : nums(nums), prefixSum(nums) {
        // Time: O(n), Space: O(n)
        int n = prefixSum.size();
        for(int i=1; i<n; i++) {
            prefixSum[i] = prefixSum[i-1] + nums[i];
        }
    }
    int sumRange(int i, int j) {
        // Time: O(1)
        if(i==0) return prefixSum[j];
        return prefixSum[j] - prefixSum[i-1];
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    NumArray *obj; vi nums; int ans;

    nums = {-2, 0, 3, -5, 2, -1};
    obj = new NumArray(nums);
    ans = obj->sumRange(0,2);
    assert(ans==1);
    ans = obj->sumRange(2,5);
    assert(ans==-1);
    ans = obj->sumRange(0,5);
    assert(ans==-3);
    delete obj;
}